### ISSUE TITLE
FOIPPA Sections Modal

### ISSUE THREAD

**Issue by m-prodan** [bcgov/foi-flow#3313]
* As an IAO users
* I want to easily select the sections of the act I am redacting under 
* so that I can quickly review and redact records

**Assumptions & Scope**	
Design for story here: https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/970284600
List of exceptions and their descriptions (FOI Requests) can be found here: https://bcgov.sharepoint.com/:x:/t/03521/EQjf_NXt0OxOiwKaYdR2u08BUwKb_KgTXnxN85r0-cjYiw?e=UfhqMe

The scope of this story covers the fact that every time IAO draws a redaction box, they need to include some kind of annotation of the section of the act they are removing it under. 

The modal should also be able to be moved, so users can view the redaction they did while selecting a code.

After a redaction is made and annotated with a code, adjusting the box with the sliders and other tools will not impact the annotations or require the user to reselect a code. 

For most requests, the annotation will just be the section number, which usually shows up inside the redaction box.
For MCFD personals - they use a "blank" redaction box, which does not use any annotation - instead, they draw one more box at the bottom of the page, and put all annotations there. 

General idea will be that once a redaction box is drawn, whether by hand or through highlighting text, the modal should appear to allow the analyst to choose the annotation. If a default code is chosen, then the modal will not appear, and will automatically apply the default code. In addition, users more commonly used FOIPPA sections will automatically reorganize to appear at the top of their modal. The more they use and redact with different sections the more the list will change.

Open Information uses a completely different set of exceptions, and will be covered in a separate user story. 

What is IN scope?
FOIPPA Redaction modal for processing team analysts

What is NOT in scope?
Redaction modal for open information/proactive disclosure severing

**Acceptance Criteria**	

**Scenario 1: Redaction Modal Menu**
* GIVEN a user is in the document review
* WHEN the user draws a redaction box
* OR highlights text for a redaction
* THEN a modal will pop out with the following design: https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/970284600 (note: click on red box to see modal)
* AND no boxes will be checked

NOTE: See link in assumptions for full list of redaction codes and their description

**Scenario 2: Selecting Redaction Sections**
* GIVEN a user is in the doc reviewer, and the FOIPPA section modal is open
* WHEN they click on the selection they want
* THEN a green check will appear in the box
* AND the user can make multiple selections 
(https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/970284603)

**Scenario 3: Removing Redaction Sections**
* GIVEN a user is in the FOIPPA section modal, and selections have been made
* WHEN they click on a green checkmark
* THEN the selection will be removed

**Scenario 4: Complete Selection**
* GIVEN a user is in the FOIPPA section modal, and selections have been made
* WHEN they click on the "Select Code(s) Button"
* THEN the modal will close
* AND the FOIPPA sections selected will appear in red text within the redaction box
(https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/970284601)

**Scenario 5: Edit Selection**
* GIVEN a user is in the  document reviewer
* WHEN they double click on a redaction box
* THEN the FOIPPA selection modal will open
* AND the sections previously selected will show in the modal

**Scenario 6: Default Codes**
* GIVEN a user is in the FOIPPA section modal, and selections have been made
* WHEN they click on save as default
* THEN the modal will close
* AND the FOIPPA sections selected will appear in red text within the redaction box

**Scenario 7: Default Codes Auto-Apply**
* GIVEN a user has set default codes in the document reviewer
* WHEN the user draws a redaction box
* OR highlights text for a redaction
* THEN the FOIPPA selection modal will NOT open
* AND the default codes will appear in the redaction box

**Scenario 8: Clear Default Codes**
* GIVEN a user has set default codes in the document reviewer
* WHEN the user double clicks on a redaction box
* THEN the FOIPPA selection modal will open
* AND the save as default button will now display as "Clear Default Codes"

**Scenario 9: Clear Default Codes**
* GIVEN a the clear default codes button is active
* WHEN the user clicks on the button
* THEN the default codes will clear
* AND any new redaction box will open the FOIPPA selection modal

**Scenario 10: Blank Redaction**
* GIVEN a user is in the FOIPPA section modal, and selected the Blank code
* WHEN they click on the "Select Code(s) Button"
* THEN the modal will close
* AND the redaction box will remain with no annotations attached to it

**Scenario 11: Cancel to Close**
* GIVEN a user is in the FOIPPA selection modal
* WHEN the user clicks on the cancel button
* THEN the modal will close
* AND the redaction box will disappear

**Scenario 12: Moving Modal**
* GIVEN a user has the FOIPPA selection modal open
* WHEN they hold click on an area of the modal that is not clickable
* AND they drag their mouse
* THEN the modal will move with their mouse cursor

**Scenario 13: Moving Modal - Release**
* GIVEN a user has the FOIPPA Selection modal open
* WHEN they have moved the modal to a different part of the screen
* AND they release their mouse click
* THEN the modal will stay in that position

**Scenario 14 - Sections Reorganize by Use**
* GIVEN a user has made multiple redactions throughout a file
* WHEN they open the FOIPPA section modal
* THEN the sections will organize, with the most used sections appearing at the top

**Scenario 15 - Sections reorganized by not using**
* GIVEN a user has had there FOIPPA selections sorted by use
* WHEN they open their modal
* THEN the sections that have not been used will appear below the sections that are being used
* AND they will be in ascending order (ie: 12, 14, 17, 18, etc...)

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/970284600
List of exceptions and description: https://bcgov.sharepoint.com/:x:/t/03521/EQjf_NXt0OxOiwKaYdR2u08BUwKb_KgTXnxN85r0-cjYiw?e=UfhqMe

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Pre-Deployment**
add REACT_APP_DOCREVIEWER_WEB_URL variable to forms flow web configmap

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by m-prodan** [comment]
FOIPPA Sections Modal story per the designs. Open for feedback @lmullane @KyEggleston @liseandtea and @JHarrietha-AOT 

Note: I made similar functionality as AXIS for default codes, I assume that is the intent we are going for.

**Comment by m-prodan** [comment]
Note from refinement: add some additional scenarios to cover adding codes for default codes

**Comment by m-prodan** [comment]
Add code organization based on useage for additional scenarios

**Comment by m-prodan** [comment]
Note: discuss at sprint planning next week with additional scenarios, is 21 still valid or should we split?

**Comment by m-prodan** [comment]
From sprint planning: for moving modal, please confirm with business/Jackie the technical solution

**Comment by m-prodan** [comment]
Per discussion at standup - need to determine re-organization, should it be re-organized at the user level, or the file level? cc @JHarrietha-AOT

**Comment by m-prodan** [comment]
From sprint planning - going to go with sections re-organized by the request level. When a user starts applying sections on a request, those sections will reorganize to the top in order of number of uses, each time the modal is refreshed/opened.

**Comment by m-prodan** [comment]
Note: @aparna-aot to take FE task from @nkan-aot as he will be away

**Comment by nkan-aot** [comment]
feb 28 standup - no double click handler for selecting redactions in pdftron so confirmed we will add a new button to popup menu to edit existing redaction
- no need for displaying selected default sections
- no way to change fill opacity in pdftron - change default redaction fill color to white, leave everything else as is

**Comment by KyEggleston** [comment]
@m-prodan @lmullane - finished testing this, and noticed a few things:

<h2>Scenario 1</h2>

The 'Delete' and 'Backspace' keys do not always delete redaction boxes. If a user does try either of these keys to delete a redaction box, they will occasionally be brought to a blank white page. I wasn't able to dig further into when or why this may happen, but it looks like it primarily happens on Excel docs, and is more likely to occur when adding redactions, closing the app, then opening it again and trying to delete. 

<h2>Scenario 10</h2>

The blank code is labelled as "s. 14 - Black Code - CFD Personals" in PDFTron. When selected, the modal will close, and the annotation will be displayed in the redaction box as "s. 14". I believe this should have no annotation and be blank? 

Also, if you have previously selected an area for redaction and selected a code from the FOIPPA modal, if you select a second area for redaction, you are able to  'Select Code(s)' without having a code selected. After clicking the  'Select Code(s)' button, the modal will close, and no annotations are placed on the redaction box (the redaction box remains). I'm not sure this is expected functionality. 

<h2>Scenario 14 and 15</h2>

It appears the section reorganization resets every time the user leaves the request in PDFTron - so if a user, leaves and comes back, it will go back to the deafult ordering (e.g., 12, 13, 14, 15 etc.) Not sure if we want this.

**Comment by m-prodan** [comment]
Good to know @KyEggleston - thanks for getting through this. 

The delete bug you observed is definitely interesting and one we might want to dig into a bit.

re: Scenario 10 - agreed, there should not be any annotation showing for this box. s. 14 would be misleading, as that is an existing exception suggesting it is legal advice.

re: 14 and 15 - likely not the behaviour we are going for. Depending on what the fix is like, we may want this addressed. In the meantime, I assume at the very least as soon as the user starts redacting again the sort will put their latest section on the top again at least?

cc @nick-aot

**Comment by nkan-aot** [comment]
@m-prodan I am working on scenario 10 14 and 15. I have not been able to reproduce scenario 1 yet so I will go back to that after I am done with the other two

**Comment by nkan-aot** [comment]
@m-prodan scenario scenario 10 14 and 15 are fixed in test. I tried downloading the excel files @KyEggleston used in his request to reproduce the bug locally but i am still unable to reproduce it, Maybe we can connect on Monday to see how it is being triggered

**Comment by KyEggleston** [comment]
@nkan-aot - was able to reproduce scenario 1 this morning, so maybe some time after stand up we can quickly go over it

**Comment by KyEggleston** [comment]
Confirmed Scen. 10, 14 and 15 updates are done, just waiting on Scen. 1 issue to be resolved (think we identified the issue)

**Comment by KyEggleston** [comment]
Resolved Scen. 1 now, closing ticket.
